tests/Unit/DataStructures/Test_DynamicBuffer.cpp

```diff
@@ -0,0 +1,119 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include "DataStructures/DynamicBuffer.hpp"
+#include "Framework/TestHelpers.hpp"
+#include "Helpers/DataStructures/MakeWithRandomValues.hpp"
+
+template <typename T>
+void check_results(const DynamicBuffer<T>& dynamic_buffer,
+                   const std::vector<T>& expected) {
+  CHECK(dynamic_buffer.size() == expected.size());
+
+  for (size_t i = 0; i < dynamic_buffer.size(); ++i) {
+    CHECK_ITERABLE_APPROX(dynamic_buffer[i], expected[i]);
+    CHECK_ITERABLE_APPROX(dynamic_buffer.at(i), expected[i]);
+  }
+
+  size_t i = 0;
+  for (auto it = dynamic_buffer.begin(); it != dynamic_buffer.end();
+       ++it, ++i) {
+    CHECK_ITERABLE_APPROX(*it, expected[i]);
+  }
+
+  i = 0;
+  for (const auto& val : dynamic_buffer) {
+    CHECK_ITERABLE_APPROX(val, expected[i]);
+    ++i;
+  }
+}
+
+template <typename T>
+std::vector<T> create_random_data(const gsl::not_null<std::mt19937*> gen,
+                                  size_t outer_size, size_t inner_size) {
+  auto value_distribution = std::uniform_real_distribution(
+      std::numeric_limits<double>::min(), std::numeric_limits<double>::max());
+
+  std::vector<T> res(outer_size);
+  for (size_t i = 0; i < outer_size; ++i) {
+    res[i] = make_with_random_values<T>(gen, value_distribution, inner_size);
+  }
+  return res;
+}
+
+template <typename T>
+void check_constructors() {
+  MAKE_GENERATOR(gen);
+  auto size_distribution = std::uniform_int_distribution(1, 10);
+  auto inner_size = static_cast<size_t>(size_distribution(gen));
+  const auto outer_size = static_cast<size_t>(size_distribution(gen));
+  if constexpr (std::is_same_v<double, T>) {
+    inner_size = 1;
+  }
+
+  CAPTURE(inner_size);
+  CAPTURE(outer_size);
+
+  const auto expected =
+      create_random_data<T>(make_not_null(&gen), outer_size, inner_size);
+
+  auto dynamic_buffer =
+      std::make_unique<DynamicBuffer<T>>(outer_size, inner_size);
+
+  for (size_t i = 0; i < outer_size; ++i) {
+    dynamic_buffer->at(i) = expected.at(i);
+  }
+
+  check_results(*dynamic_buffer, expected);
+
+  DynamicBuffer<T> dynamic_buffer_copied(*dynamic_buffer);
+  DynamicBuffer<T> dynamic_buffer_assigned = *dynamic_buffer;
+
+  if constexpr (tt::is_iterable_v<T>) {
+    // check that the new buffers don't point to the same data storage location
+
+    // first data_vector of dynamic_buffer
+    const auto first_data_vector = *dynamic_buffer->begin();
+
+    // underlying C array
+    const auto array = first_data_vector.data();
+
+    // address of first element
+    const auto address = &*array;
+
+    // analogously for copied DynamicBuffers
+    CHECK(&*((*dynamic_buffer_copied.begin()).data()) != address);
```

Comment:
can you verify this for _all_ `DataVector`s. It's easy to write code that works for only the first or last element but fails for things in the middle :) (I don't think that's the case here, but it's good to be sure)